### PR TITLE
Editing test_response.py and adding new notebook

### LINKED ISSUE #2: Notebook with code to make detector summed responses
Here's the most recent notebook I have to make detector summed responses that I used for the upper limits. Zipped because github doesn't allow .ipynb files 
[MakingDRMsFromNITRATESresps-Copy2.zip](https://github.com/Swift-BAT/UnmaskedResponder/files/11872386/MakingDRMsFromNITRATESresps-Copy2.zip)

### PR DESCRIPTION
**Pull request by metalphase**
Adding notebook for calculating responses. Adding resp_tabs to config.py. Revising test_response.py to use alternative method of calculating response.

### CODE REVIEW

**Comment by metalphase** [comment]
@Tohuvavohu I compared the sum of the nitrates response dpi with the sum of the response dpi created by the new package. Just to clear up what I'm doing here. I've included nitrates_total_resp.npy which is the summed response dpi for the same model generated using the nitrates package. I generate the same summed response dpi using the UnmaskedResponder package and use numpy.allclose to make sure that the values across these arrays are within a relative tolerance of 1e-08 of each other. 

I've elected to use the sums as opposed to the base dpis since the .npy file for the dpi generated in nitrates was around 450mb and is greater than the size limit imposed by git.

**Comment by metalphase** [comment]
I've also noticed that in config.py we use RESP_TAB_DNAME for the  resp_tabs_ebins directory however we didn't have anything for the resp_tabs directory. I've added an entry for the path for the resp_tabs directory in config.py. This was required in order to create the signal model which required resp_tab_dname as an argument

**Comment by Tohuvavohu** [comment]
Awesome @metalphase ! Looks really great :) 

Three things:

1. Can you please add in the basic response files needed for the test to run locally (ie for just the one position...without all of the responses installed) like you did for the tests in the NITRATES package? Then we can have the test run automatically on each PR via Github Actions.
2. Can you add a basic function to the package, that takes theta,phi and produces a complete response? ie wrap a bunch of the code in the notebook into a utility function of the package?
3. Can you add another function that will take that complete response, sum it, and then save the summed responses in the format .rsp file as shown in Cell 13 of Jimmy's notebook?


Thanks!

**Comment by metalphase** [comment]
@Tohuvavohu Added the same files in nitrates_resp_dir into the tests directory for the package. I also added tools.source_model_response.py which includes generate_response_dpis (generates the dpis) and save_summed_response (saves the responses into a .rsps file and saves into a new directory /signal_rsps).

**Comment by Tohuvavohu** [comment]
@metalphase Thanks!
Two things:

1) There is no reason to duplicate the same code tha is in generate_response_dpis inside save_summed_response. Call generate from within summed reponse. Also the name 'save_summed_response' is not great, just call it summed_reponse.
Also, think about these functions from the perspective of a user, don't save to a random directory, have the outdirectory be a parameter option, with the default being the local dir.

2) Thanks for moving things over!
The goal here is that the pytest should run natively without the response directories set up (sourcing only from the local package).
Right now i get:
`tests/test_response.py F                                                 [100%]

=================================== FAILURES ===================================
___________________________ test_response_generation ___________________________

    def test_response_generation():
    
        import nitrates_response as nr
    
        ebins0 = np.array([15.0, 24.0, 35.0, 48.0, 64.0])
        ebins0 = np.append(ebins0, np.logspace(np.log10(84.0), np.log10(500.0), 5+1))[:-1]
        ebins0 = np.round(ebins0, decimals=1)[:-1]
        ebins1 = np.append(ebins0[1:], [350.0])
        ebins = [ebins0, ebins1]
    
        th = 0.
        ph = 0.
    
        flux_mod = nr.models.flux_models.Cutoff_Plaw_Flux(E0=100.0)
        bl_alldets = get_bldmask_alldets()
>       rt_obj = nr.response.ray_trace_funcs.RayTraces(nr.config.rt_dir)

tests/test_response.py:43: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/nitrates_response/response/ray_trace_funcs.py:473: in __init__
    self.rt_arr = get_rt_arr(rt_dir, ident=ident)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

rt_dir = '../ray_traces_detapp_npy', ident = 'fwd_ray_trace'

    def get_rt_arr(rt_dir, ident="fwd_ray_trace"):
>       ray_trace_fnames = np.array([fn for fn in os.listdir(rt_dir) if ident in fn])
E       FileNotFoundError: [Errno 2] No such file or directory: '../ray_traces_detapp_npy'

src/nitrates_response/response/ray_trace_funcs.py:12: FileNotFoundError
`

File: tests/nitrates_resp_dir/bkg_estimation.csv
Comment (Tohuvavohu) [inline tests/nitrates_resp_dir/bkg_estimation.csv:1]: you dont need this file.

File: tests/nitrates_resp_dir/bright_src_cat.fits
Comment (Tohuvavohu) [inline tests/nitrates_resp_dir/bright_src_cat.fits:1]: you don't need this file

File: tests/nitrates_resp_dir/comp_flor_resps/.DS_Store
Comment (Tohuvavohu) [inline tests/nitrates_resp_dir/comp_flor_resps/.DS_Store:1]: remove

File: tests/nitrates_resp_dir/filter_evdata.fits
Comment (Tohuvavohu) [inline tests/nitrates_resp_dir/filter_evdata.fits:1]: remove

File: tests/nitrates_resp_dir/results.db
Comment (Tohuvavohu) [inline tests/nitrates_resp_dir/results.db:1]: remove

**Comment by metalphase** [comment]
@Tohuvavohu I was able to get save_summed_response to generate the file from the dpis (I previously reused the same code since the tables referred to objects that were only accessible within the generate_response_dpis). Somehow the test_response.py is no longer passing tests so I'm trying to figure out how it's occuring before commiting

**Comment by metalphase** [comment]
@Tohuvavohu still investigating the assertionError. I am seeing different results after going through the same exact process detailed in test_response_generation() using the UnmaskedResponder package and the nitrates package separate.

I've updated my nitrates repo locally and re-calculated the values for nitrates_total_resp.npy so this should be up-to-date and not the source of the issues. I suspect some of the response files may have been renamed perhaps. I'll keep you updated

**Comment by Tohuvavohu** [comment]
None of the response files have been changed on Zenodo, so if you expect a problem, you should redownload them.
How were the tests passing before?

**Comment by metalphase** [comment]
@Tohuvavohu finally resolved it. It was definitely the response files on my end. Running pytest on test_response.py is successful on my end. 

The generate_response_dpis function now returns PhotonEmins and PhotonEmaxs so that summed_response can refer to these when constructing the dictionary for the table.

**Comment by metalphase** [comment]
Just cloned a new copy of the repo and ran pytest on test_response.py and it worked for me. Let me know if you have any issues!

**Comment by Tohuvavohu** [comment]
Looking good @metalphase!
I see you changed the relative tolerance on the test from 1e-8 to 1e-6. Was there a reason for this?

Otherwise, I think just need a squash (because lots of files came in/out, and don't want to carry them around in commit history). And then we can merge.

**Comment by metalphase** [comment]
ah, I changed the relative tolerance for testing purposes to investigate the assertion errors I was having. Will update it back to the original 1e-08.

**Comment by metalphase** [comment]
Commits should be squashed and ready to be merged

**Pull request merged**